cmd/slowlog-faker/main.go

```diff
@@ -0,0 +1,543 @@
+			// e.Server = fmt.Sprintf("db%v", zipf9.Uint64()+1)       // fake data
+			// e.Host = fmt.Sprintf("10.11.12.%v", zipf99.Uint64()+1) // fake data
+
+			e.Db = fmt.Sprintf("schema%d", r.Intn(100))      // fake data 100
+			e.User = fmt.Sprintf("user%d", r.Intn(100))      // fake data 100
+			e.Host = fmt.Sprintf("10.11.12.%d", r.Intn(100)) // fake data 100
+			e.Server = fmt.Sprintf("db%d", r.Intn(10))       // fake data 10
+			e.LabelsKey = []string{fmt.Sprintf("label%d", r.Intn(10)), fmt.Sprintf("label%d", r.Intn(10)), fmt.Sprintf("label%d", r.Intn(10))}
+			e.LabelsValue = []string{fmt.Sprintf("value%d", r.Intn(100)), fmt.Sprintf("value%d", r.Intn(100)), fmt.Sprintf("value%d", r.Intn(100))}
+
+			aggregator.AddEvent(e, digest, e.User, e.Host, e.Db, e.Server, fingerprint)
+
+			// Pass last offset to restart reader when reached out end of slowlog.
+			opt.StartOffset = e.OffsetEnd
+
+			i++
+
+			if periodStart.IsZero() {
+				periodStart = logStart.Add(time.Duration(periodNumber) * time.Minute)
+			}
+
+			if prewTs.IsZero() {
+				prewTs = e.Ts
+			}
+
+			if e.Ts.Sub(prewTs).Seconds() > 59 {
+				prewTs = e.Ts
+				break
+			}
+		}
+		periodNumber++
+
+		// No new events in slowlog. Nothing to save in ClickHouse. New iteration.
+		if i == 0 {
+			fmt.Printf("Iteration %d of %d \n", iteration, *repeatN)
+			if iteration > *repeatN {
+				fmt.Printf("Done. Total Iterations %v \n", iteration)
+				break
+			}
+			iteration++
+			opt.StartOffset = 0
+			events = parseSlowLog(*slowLogPath, opt)
+			continue
+		}
+
+		res := aggregator.Finalize()
+
+		j := 0
+		classesLen := len(res.Class) - 1
+		var stmt *sql.Stmt
+		var tx *sql.Tx
+		for _, v := range res.Class {
+
+			// If key has suffix _time or _wait than field is TimeMetrics.
+			// For Boolean metrics exists only Sum.
+			// https://www.percona.com/doc/percona-server/5.7/diagnostics/slow_extended.html
+			// TimeMetrics: query_time, lock_time, rows_sent, innodb_io_r_wait, innodb_rec_lock_wait, innodb_queue_wait.
+			// NumberMetrics: rows_examined, rows_affected, rows_read, merge_passes, innodb_io_r_ops, innodb_io_r_bytes,
+			// innodb_pages_distinct, query_length, bytes_sent, tmp_tables, tmp_disk_tables, tmp_table_sizes.
+			// BooleanMetrics: qc_hit, full_scan, full_join, tmp_table, tmp_table_on_disk, filesort, filesort_on_disk,
+			// select_full_range_join, select_range, select_range_check, sort_range, sort_rows, sort_scan,
+			// no_index_used, no_good_index_used.
+
+			if j%100000 == 0 {
+				tx, err = connect.Begin()
+				if err != nil {
+					fmt.Printf("transaction begin error: %v", err)
+				}
+				stmt, err = tx.Prepare(insertSQL)
+				if err != nil {
+					fmt.Printf("prepare error: %v", err)
+				}
+			}
+			numQueries := float32(r.Intn(10000) + 1)
+			args := makeValues(v, periodStart, numQueries)
+			_, err = stmt.Exec(args...)
+			if err != nil {
+				fmt.Printf("exec error: %v", err)
+			}
+
+			if j >= classesLen || j%100000 == 99999 {
```

Comment:
I think this is kind of weird batching.

It can be simplified by extracting the whole Begin/Prepare/Exec/Close/Commit code into a function accepting a slice of classes. Split aggregation results into slices, call that function for every slice.

Bonus point - each call can be in its own goroutine. That should improve insertion speed: https://clickhouse.yandex/docs/en/introduction/performance/#performance-when-inserting-data

Another bonus point – `defer`s can be used.